integreat_cms/summ_ai_api/utils.py

```diff
@@ -392,3 +416,169 @@ def __repr__(self):
+
+        :returns: a task of the queue
+        :rtype: asyncio.Future
+        """
+        now = time.time()
+        if (
+            self.last_rate_limit is not None
+            and now - self.last_rate_limit < self.wait_time
+        ):
+            # If we are currently waiting out a rate limit,
+            # the worker should sleep for the remaining time before getting the next real task.
+            # Return a function without arguments that the worker can evaluate
+            # to obtain a coroutine.
+            # (see the corresponding comment in SummAiApiClient…translate_text_fields())
+            return delay_function_call(
+                asyncio.sleep, [self.wait_time - (now - self.last_rate_limit)]
+            )
+        if len(self) == 0:
+            raise StopIteration()
+        return self.popleft()
+
+    def hit_rate_limit(self, task):
+        """
+        A task hit the rate limit, so wait a bit and reschedule the task
+
+        :param task: The task that failed because of the rate limiting
+        :type task: asyncio.Future
+        """
+
+        # Only save current timestamp if this is the first failed request reported
+        if (
+            self.last_rate_limit is None
+            or time.time() - self.last_rate_limit < self.wait_time
+        ):
+            self.last_rate_limit = time.time()
+        # Reschedule the failed task
+        self.appendleft(task)
+
+
+async def worker(loop, task_generator, i=None):
+    """
+    Continuously gets a task from the queue and executes it.
+    Stops once no more tasks are available.
+    This form makes it easy to always have at most n concurrent tasks
+    as well as intermittent wait times through the task generator.
+
+    :param loop: The asyncio event loop to execute tasks in
+    :type loop: asyncio.AbstractEventLoop
+
+    :param task_generator: Queue to execute tasks from
+    :type task_generator: PatientTaskQueue
+
+    :param i: ID of the worker (for logging purposes)
+    :type i: str
+
+    :returns: A list of task-results
+    :rtype: List
+    """
+    logger.debug("Worker #%s initialized", i)
+
+    completed = []
+    while len(task_generator):
+        task = next(task_generator)
+        try:
+            result = await loop.create_task(task())
+        except SummAiRateLimitingExceeded:
+            task_generator.hit_rate_limit(task)
+        else:
+            completed.append(result)
+
+    logger.debug("Worker #%s completed %s tasks", i, len(completed))
+    return completed
+
+
+def flattened_list(l):
+    """
+    Helper method to smooth nested lists
+
+    :param l: A (maybe nested) list
+    :type l: list
+
+    :return: a flattened list
+    :rtype: list
+    """
+
+    result = []
+    for x in l:
+        if isinstance(x, list):
+            result.extend(flattened_list(x))
+        else:
+            result.append(x)
+    return result
+
+
+def delay_function_call(func, args=None):
```

Comment:
Just for the future: When you're working with a dynamic number of arguments (which can also be no arguments at all), it's often time easier to work with the asterisk notation `*args` in the function signature (see [Arbitrary Argument Lists](https://docs.python.org/3/tutorial/controlflow.html#arbitrary-argument-lists)).